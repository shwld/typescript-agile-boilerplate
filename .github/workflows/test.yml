name: test

on:
  push:
    branches: ["main"]
  pull_request:
    types: [opened, synchronize]

jobs:
  build:
    name: Test
    timeout-minutes: 15
    runs-on: ${{ matrix.os }}
    # To use Remote Caching, uncomment the next lines and follow the steps below.
    # env:
    #  TURBO_TOKEN: ${{ secrets.TURBO_TOKEN }}
    #  TURBO_TEAM: ${{ secrets.TURBO_TEAM }}
    strategy:
      matrix:
        # os: [ubuntu-latest, macos-latest]
        os: [ubuntu-latest]
    services:
      # Label used to access the service container
      postgres:
        # Docker Hub image
        image: postgres
        # Provide the password for postgres
        env:
          POSTGRES_USER: postgres
          POSTGRES_PASSWORD: password
          POSTGRES_DB: app_test
        # Set health checks to wait until postgres has started
        options: >-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5

    steps:
      - name: Check out code
        uses: actions/checkout@v2
        with:
          fetch-depth: 2

      - name: Setup Node.js environment
        uses: actions/setup-node@v2
        with:
          node-version: 16
          cache: 'yarn'

      - name: Setup Test
        run: |
          echo 'DATABASE_URL="postgresql://postgres:password@postgres:5432/app_test?schema=public"' > ./infrastructures/db/.env
          echo 'DATABASE_URL="postgresql://postgres:password@postgres:5432/app_test?schema=public"' > ./use-cases/graphql-resolvers/.env
          yarn db:migrate

      - name: Install dependencies
        run: yarn

      - name: Test
        run: yarn test
