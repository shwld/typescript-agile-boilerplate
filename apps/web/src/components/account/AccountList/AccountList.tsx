import { Button, ListItem, UnorderedList } from '@chakra-ui/react';
import { ReactNode, FC, useState } from 'react';
import { AccountUpdateForm } from '../AccountUpdateForm';
import { useAccountListQuery } from './AccountList.generated';

export const AccountList: FC<{ children?: ReactNode }> = () => {
  const [cursor, setCursor] = useState('');
  const [result] = useAccountListQuery({ variables: { cursor } });

  const accounts =
    result.data?.viewer?.accounts.edges?.map((it) => it?.node!) ?? [];
  const { hasNextPage, endCursor } =
    result.data?.viewer?.accounts.pageInfo ?? {};

  return (
    <UnorderedList>
      {accounts.map((account) => (
        <ListItem key={account.id}>
          <AccountUpdateForm
            defaultValues={{ id: account.id, name: account.name }}
          />
        </ListItem>
      ))}
      {hasNextPage && (
        <Button
          onClick={() => {
            endCursor != null && setCursor(endCursor);
          }}
        >
          More
        </Button>
      )}
    </UnorderedList>
  );
};
