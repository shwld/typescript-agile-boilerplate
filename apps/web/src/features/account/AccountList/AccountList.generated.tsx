import * as Types from '../../../graphql/generated/graphql';

import gql from 'graphql-tag';
import * as Urql from 'urql';
export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
export type AccountListResultFragment = { __typename?: 'Account', id: string, name: string };

export type AccountListQueryVariables = Types.Exact<{
  cursor?: Types.InputMaybe<Types.Scalars['String']>;
}>;


export type AccountListQuery = { __typename?: 'Query', viewer?: { __typename?: 'Viewer', id: string, accounts: { __typename?: 'AccountConnection', edges?: Array<{ __typename?: 'AccountEdge', cursor?: string | null, node?: { __typename?: 'Account', id: string, name: string } | null } | null> | null, pageInfo?: { __typename?: 'PageInfo', hasNextPage: boolean, endCursor?: string | null } | null } } | null };

export const AccountListResultFragmentDoc = gql`
    fragment AccountListResult on Account {
  id
  name
}
    `;
export const AccountListDocument = gql`
    query accountList($cursor: String) {
  viewer {
    id
    accounts(first: 3, after: $cursor) {
      edges {
        node {
          ...AccountListResult
        }
        cursor
      }
      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
}
    ${AccountListResultFragmentDoc}`;

export function useAccountListQuery(options?: Omit<Urql.UseQueryArgs<AccountListQueryVariables>, 'query'>) {
  return Urql.useQuery<AccountListQuery>({ query: AccountListDocument, ...options });
};