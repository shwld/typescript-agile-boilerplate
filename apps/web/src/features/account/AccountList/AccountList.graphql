fragment AccountList_Result on Account {
  id
  name
  projects {
    edges {
      node {
        id
        name
      }
      cursor
    }
    pageInfo {
      hasNextPage
      endCursor
    }
  }
}

query AccountList($cursor: String) {
  viewer {
    id
    accounts(first: 10, after: $cursor) {
      edges {
        node {
          ...AccountList_Result
        }
        cursor
      }
      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
}
