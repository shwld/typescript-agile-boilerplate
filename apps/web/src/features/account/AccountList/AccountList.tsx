import { Button, Box, Flex, UnorderedList, Text } from '@chakra-ui/react';
import { ReactNode, FC, useState, Fragment } from 'react';
import { ProjectCreateButton } from '~/features/project/ProjectCreateButton';
import { AccountUpdateButton } from '../AccountUpdateButton';
import { useAccountListQuery } from './AccountList.generated';
import Link from 'next/link';

export const AccountList: FC<{ children?: ReactNode }> = () => {
  const [cursor, setCursor] = useState('');
  const [result] = useAccountListQuery({ variables: { cursor } });

  const accounts =
    result.data?.viewer?.accounts.edges?.map((it) => it?.node!) ?? [];
  const { hasNextPage, endCursor } =
    result.data?.viewer?.accounts.pageInfo ?? {};

  return (
    <UnorderedList aria-label="Account list">
      {accounts.map((account) => (
        <Box key={account.id} borderWidth="1px">
          <Box bgColor="black" p={3}>
            <Flex justify="space-between" alignItems="center">
              <Text color="white">{account.name}</Text>
              <AccountUpdateButton
                defaultValues={{ id: account.id, name: account.name }}
              />
            </Flex>
          </Box>
          {account.projects.edges?.map((project) => (
            <Fragment key={project?.node?.id}>
              {project?.node != null && (
                <Box bgColor="gray" p={3}>
                  <Flex justify="space-between" alignItems="center">
                    <Link href={`/projects/${project.node.id}`}>
                      <a>
                        <Text>{project.node.name}</Text>
                      </a>
                    </Link>
                  </Flex>
                </Box>
              )}
            </Fragment>
          ))}
          <ProjectCreateButton accountId={account.id} />
        </Box>
      ))}
      {hasNextPage && (
        <Button
          onClick={() => {
            endCursor != null && setCursor(endCursor);
          }}
        >
          More
        </Button>
      )}
    </UnorderedList>
  );
};
