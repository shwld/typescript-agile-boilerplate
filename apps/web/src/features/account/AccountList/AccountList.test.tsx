import { AccountList } from './AccountList';
import { render } from '@testing-library/react';
import { MockedUrqlProvider } from '~/test/MockedUrqlProvider';

describe('AccountList', () => {
  const renderComponent = () => {
    const renderResult = render(
      <MockedUrqlProvider>
        <AccountList />
      </MockedUrqlProvider>
    );
    return renderResult;
  };
  test('Snapshot', () => {
    expect(renderComponent().asFragment()).toMatchSnapshot();
  });
  test('success', () => {
    const { getByLabelText } = renderComponent();
    expect(getByLabelText('Account list')).toBeTruthy();
  });
});
