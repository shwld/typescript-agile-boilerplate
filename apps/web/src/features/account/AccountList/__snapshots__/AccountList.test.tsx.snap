// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountList Snapshot 1`] = `
<DocumentFragment>
  <ul
    aria-label="Account list"
    class="css-tu0njr"
    role="list"
  >
    <div
      class="css-sajjl8"
    >
      <div
        class="css-17i2t5k"
      >
        <div
          class="css-1lekzkb"
        >
          <p
            class="chakra-text css-kbn7if"
          >
            ut
          </p>
          <button
            class="chakra-button css-5c2176"
            type="button"
          >
            EDIT ACCOUNT
          </button>
        </div>
      </div>
      <div
        class="css-bpjgaa"
      >
        <div
          class="css-1lekzkb"
        >
          <a
            href="/projects/3e7d7963-406a-47bd-ad80-0448296895bf"
          >
            <p
              class="chakra-text css-0"
            >
              et
            </p>
          </a>
        </div>
      </div>
      <button
        class="chakra-button css-5c2176"
        type="button"
      >
        CREATE PROJECT
      </button>
    </div>
    <button
      class="chakra-button css-5c2176"
      type="button"
    >
      More
    </button>
  </ul>
</DocumentFragment>
`;
